service/history/ndc/workflow_resetter_test.go

```diff
@@ -972,3 +970,106 @@ func (s *workflowResetterSuite) TestPagination() {
+		ctx,
+		gomock.Any(),
+		definition.NewWorkflowKey(s.namespaceID.String(), s.workflowID, s.baseRunID),
+		baseBranchToken,
+		baseRebuildLastEventID,
+		util.Ptr(baseRebuildLastEventVersion),
+		definition.NewWorkflowKey(s.namespaceID.String(), s.workflowID, s.resetRunID),
+		resetBranchToken,
+		resetRequestID,
+	).Return(resetMutableState, resetHistorySize, nil)
+
+	resetMutableState.EXPECT().SetBaseWorkflow(s.baseRunID, baseRebuildLastEventID, baseRebuildLastEventVersion)
+	resetMutableState.EXPECT().AddHistorySize(resetHistorySize)
+	resetMutableState.EXPECT().GetCurrentVersion().Return(resetWorkflowVersion).AnyTimes()
+	resetMutableState.EXPECT().UpdateCurrentVersion(resetWorkflowVersion, false).Return(nil).AnyTimes()
+	resetMutableState.EXPECT().GetExecutionState().Return(&persistencespb.WorkflowExecutionState{
+		RunId:  resetRequestID,
+		Status: enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING,
+	}).AnyTimes()
+	resetMutableState.EXPECT().GetExecutionInfo().Return(&persistencespb.WorkflowExecutionInfo{}).AnyTimes()
+	resetMutableState.EXPECT().GetPendingActivityInfos().Return(map[int64]*persistencespb.ActivityInfo{})
+	resetMutableState.EXPECT().RefreshExpirationTimeoutTask(ctx).Return(nil).AnyTimes()
+	resetMutableState.EXPECT().GetPendingChildExecutionInfos().Return(executionInfos)
+
+	workflowTaskSchedule := &workflow.WorkflowTaskInfo{
+		ScheduledEventID: baseRebuildLastEventID - 12,
+		StartedEventID:   common.EmptyEventID,
+		RequestID:        uuid.New(),
+		TaskQueue: &taskqueuepb.TaskQueue{
+			Name: "random task queue name",
+			Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+		},
+	}
+	resetMutableState.EXPECT().GetPendingWorkflowTask().Return(workflowTaskSchedule).AnyTimes()
+
+	workflowTaskStart := &workflow.WorkflowTaskInfo{
+		ScheduledEventID: workflowTaskSchedule.ScheduledEventID,
+		StartedEventID:   workflowTaskSchedule.ScheduledEventID + 1,
+		RequestID:        workflowTaskSchedule.RequestID,
+		TaskQueue:        workflowTaskSchedule.TaskQueue,
+	}
+	resetMutableState.EXPECT().AddWorkflowTaskStartedEvent(
+		workflowTaskSchedule.ScheduledEventID,
+		workflowTaskSchedule.RequestID,
+		workflowTaskSchedule.TaskQueue,
+		consts.IdentityHistoryService,
+		nil,
+		nil,
+	).Return(&historypb.HistoryEvent{}, workflowTaskStart, nil)
+	resetMutableState.EXPECT().AddWorkflowTaskFailedEvent(
+		workflowTaskStart,
+		enumspb.WORKFLOW_TASK_FAILED_CAUSE_RESET_WORKFLOW,
+		failure.NewResetWorkflowFailure(resetReason, nil),
+		consts.IdentityHistoryService,
+		nil,
+		"",
+		s.baseRunID,
+		s.resetRunID,
+		baseRebuildLastEventVersion,
+	).Return(&historypb.HistoryEvent{}, nil)
+
+	resetWorkflow, err := s.workflowResetter.prepareResetWorkflow(
+		ctx,
+		s.namespaceID,
+		s.workflowID,
+		s.baseRunID,
+		baseBranchToken,
+		baseRebuildLastEventID,
+		baseRebuildLastEventVersion,
+		s.resetRunID,
+		resetRequestID,
+		resetWorkflowVersion,
+		resetReason,
+	)
+	s.NoError(err)
+	s.Equal(resetMutableState, resetWorkflow.GetMutableState())
```

Comment:
We will definitely need a test to make sure reset workflow is timed out properly. That is a timeout task is generated and it is fired as expected.

Reply:
This test is checking that timeout was actually generated.
I don't think that we need a test that waits that timeout is actually fired.
Also all of that is kind of not related - reason why workflow was instantly timing out is fixed and covered by unit test.

Reply:
But the unit test is not covering RefreshTasksForWorkflowStart() right?

Reply:
unit test is covering the code that was causing the problem, which is not RefreshTasksForWorkflowStart.
this test is making sure RefreshTasksForWorkflowStart is called.